src/contrib/ethosu/cascader/proposal_generator.cc

```diff
@@ -0,0 +1,219 @@
+  // non-local memory that's required in the Proposal at that point in time
+  int new_memory_usage = plan->GetMemoryUsage();
+  for (const auto& it : new_configs) {
+    if (plan_boundary_configs.find(it.second) == plan_boundary_configs.end()) {
+      new_memory_usage += it.first->GetSize();
+    }
+  }
+  new_memory_usage = std::max(new_memory_usage, proposal->GetMemoryUsage());
+  int new_cycles = proposal->GetCycles() + plan->GetCycles();
+  std::vector<Part> new_part_group = proposal->GetPartGroup();
+  new_part_group.insert(new_part_group.end(), plan->GetPartGroup().begin(),
+                        plan->GetPartGroup().end());
+  std::sort(new_part_group.begin(), new_part_group.end());
+  return Proposal(proposal->GetGraph(), new_part_group, new_plans, new_configs,
+                  proposal->GetCascadeRegion(), new_memory_usage, new_cycles);
+}
+
+std::vector<Proposal> GeneratePartialProposals(
+    const CascaderGraph& graph, const HomeMap& home_map, const CascaderOptions options,
+    const std::unordered_map<Part, std::vector<Plan>, ObjectPtrHash, ObjectPtrEqual>& plans_by_part,
+    const std::vector<Part>& partial_proposal_group,
+    std::unordered_map<std::vector<Part>, std::vector<Proposal>>* proposals_by_group) {
+  if (proposals_by_group->find(partial_proposal_group) != proposals_by_group->end()) {
+    return proposals_by_group->at(partial_proposal_group);
+  }
+  if (partial_proposal_group.size() == 0) {
+    (*proposals_by_group)[partial_proposal_group] =
+        std::vector<Proposal>{Proposal(graph, std::vector<Part>(), std::vector<Plan>(),
+                                       TensorConfigMap(), options->cascade_region, 0, 0)};
+  } else {
+    Part part = partial_proposal_group.back();
+    const auto& plans = plans_by_part.at(part);
+    for (const auto& plan : plans) {
+      if (plan->GetInteriorRegion() == options->cascade_region) {
+        // Doing this isn't very efficient, but it improves the performance of the Plan
+        // generator
+        std::unordered_set<TensorConfig> plan_boundary_configs = GetPlanBoundaryConfigs(plan);
+        // The residual_proposal_group is a Part group indicating the Parts which aren't
+        // covered by the current Plan. It's the group for which we must find 'residual
+        // Proposals', meaning Proposals which cover the rest of the CascaderGraph assuming we
+        // pick the current Plan.
+        std::vector<Part> residual_proposal_group;
+        std::copy_if(partial_proposal_group.begin(), partial_proposal_group.end(),
+                     std::back_inserter(residual_proposal_group), [&plan](Part value) {
+                       return std::find(plan->GetPartGroup().begin(), plan->GetPartGroup().end(),
+                                        value) == plan->GetPartGroup().end();
+                     });
+        // std::sort(residual_proposal_group.begin(), residual_proposal_group.end());
+        const auto& residual_proposals = GeneratePartialProposals(
```

Comment:
A comment would help here to differentiate between "residual_proposal_group" vs "residual_proposals"